packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -1432,38 +1363,136 @@ class TextPainter {
   ///
   /// Valid only after [layout] has been called.
   double? getFullHeightForCaret(TextPosition position, Rect caretPrototype) {
-    if (position.offset < 0) {
-      // TODO(LongCatIsLooong): make this case impossible; see https://github.com/flutter/flutter/issues/79495
-      return null;
-    }
-    return switch (_computeCaretMetrics(position)) {
-      _LineCaretMetrics(:final double fullHeight) => fullHeight,
-      _EmptyLineCaretMetrics() => null,
-    };
+    final TextBox textBox = _getOrCreateLayoutTemplate().getBoxesForRange(0, 1, boxHeightStyle: ui.BoxHeightStyle.strut).single;
+    return textBox.toRect().height;
   }
-
+  bool _isNewlineAtOffset(int offset) => 0 <= offset && offset < plainText.length
+                                      && WordBoundary._isNewline(plainText.codeUnitAt(offset));
   // Cached caret metrics. This allows multiple invokes of [getOffsetForCaret] and
   // [getFullHeightForCaret] in a row without performing redundant and expensive
   // get rect calls to the paragraph.
-  late _CaretMetrics _caretMetrics;
+  //
+  // The cache implementation assumes there's only one cursor at any given time.
+  late _LineCaretMetrics _caretMetrics;
 
-  // Checks if the [position] and [caretPrototype] have changed from the cached
-  // version and recomputes the metrics required to position the caret.
-  _CaretMetrics _computeCaretMetrics(TextPosition position) {
+  // This function returns the caret's offset and height for the given
+  // `position` in the text, or null if the paragraph is empty.
+  //
+  // For a TextPosition, typically when its TextAffinity is downstream, the
+  // corresponding I-beam caret is anchored to the leading edge of the
+  // `offset`-th character in the text. When the TextAffinity is upstream, the
```

Comment:
Maybe: "`offset`-th character" => "character at `offset`"

So for offset 1 and affinity downstream, it's anchored to the leading edge of the 2nd character (the character at index 1), right? If so, I think the above change is more clear. When I first read this paragraph, I think I incorrectly interpreted it to mean the 1st character.

Reply:
That's right.